### PR TITLE
[Merged by Bors] - feat: port SetTheory.Ordinal.Notation

### PR DESCRIPTION
**Pull request by mo271**
---

- [x] depends on: #4281

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by joneugster** [comment]
I just rebased this to the newest master, but I don't actually manage to make any progress :(

**Comment by vihdzp** [comment]
I'll make an attempt to clean up the Mathlib 3 file. Hopefully that helps porting this file.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#4281~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by Ruben-VandeVelde** [comment]
> I'll make an attempt to clean up the Mathlib 3 file. Hopefully that helps porting this file.

Any news, @vihdzp ?

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 1]
```diff
@@ -0,0 +1,1384 @@
...
+set_option linter.uppercaseLean3 false
+
+
+open Ordinal Order
+
+open Ordinal
+
+-- Porting note: the generated theorem gets lint.
+set_option genSizeOfSpec false in
+-- get notation for `ω`
```
Comment (semorrison) [inline Mathlib/SetTheory/Ordinal/Notation.lean:38]: Was this comment misplaced by mathport? Perhaps check against the mathilb3 source.

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 2]
```diff
@@ -0,0 +1,1384 @@
...
+@[simp]
+theorem repr_ofNat (n : ℕ) : repr (ofNat n) = n := by cases n <;> simp
+#align onote.repr_of_nat ONote.repr_ofNat
+
+-- @[simp] -- Porting note: simp can prove this
+theorem repr_one : repr (ofNat 1) = (1 : ℕ) := repr_ofNat 1
+#align onote.repr_one ONote.repr_one
+
+theorem omega_le_oadd (e n a) : ω ^ repr e ≤ repr (oadd e n a) := by
+  --unfold repr
```
Comment (semorrison) [inline Mathlib/SetTheory/Ordinal/Notation.lean:165]: Remove?

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 3]
```diff
@@ -0,0 +1,1384 @@
...
+    rw [← add_one_eq_succ, add_mul_succ _ (one_add_of_omega_le h), add_one_eq_succ, succ_le_iff,
+      Ordinal.mul_lt_mul_iff_left (Ordinal.pos_iff_ne_zero.2 e0)]
+    exact omega_isLimit.2 _ l
+  · apply (principal_mul_omega (omega_isLimit.2 _ h) l).le.trans
+    simpa using mul_le_mul_right' (one_le_iff_ne_zero.2 e0) ω
+#align onote.repr_opow_aux₁ ONote.repr_opow_aux₁
+
+section
+
+-- Porting note: `R'` is used many times but it gets lint.
```
Comment (semorrison) [inline Mathlib/SetTheory/Ordinal/Notation.lean:890]: Could you add more context here? I'm not sure what this means.

File: Mathlib/SetTheory/Ordinal/Notation.lean [hunk 4]
```diff
@@ -0,0 +1,1384 @@
...
+    | Sum.inl (some a), h =>
+      have : a < o := by rw [lt_def, h.1]; apply lt_succ
+      fun i => (fastGrowing a^[i]) i
+    | Sum.inr f, h => fun i =>
+      have : f i < o := (h.2.1 i).2.1
+      fastGrowing (f i) i
+  termination_by fastGrowing o => o
+#align onote.fast_growing ONote.fastGrowing
+
+-- Porting note: `this` in the definition of `fastGrowing` gets lint.
```
Comment (semorrison) [inline Mathlib/SetTheory/Ordinal/Notation.lean:1166]: Sorry, don't understand "gets lint". I assume this is failing one of the linters?

File: Mathlib/SetTheory/Ordinal/Notation.lean (hunk 1 above)
Comment (joneugster) [inline Mathlib/SetTheory/Ordinal/Notation.lean:38]:
```suggestion
open Ordinal Order

-- Porting note: the generated theorem gets lint.
set_option genSizeOfSpec false in
```

The mathlib3 source had
```
open ordinal order
open_locale ordinal -- get notation for `ω`
```
that's where the comment  comes from

Comment (joneugster) [reply Mathlib/SetTheory/Ordinal/Notation.lean:38]: see code suggestion below

File: Mathlib/SetTheory/Ordinal/Notation.lean (hunk 4 above)
Comment (semorrison) [reply Mathlib/SetTheory/Ordinal/Notation.lean]: There are still comments that say "gets lint". Could we fix those prior to merging?

File: Mathlib/SetTheory/Ordinal/Notation.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/SetTheory/Ordinal/Notation.lean]: It seems the comment has now vanished from the source. Could it be restored, please?

Comment (Komyyy) [reply Mathlib/SetTheory/Ordinal/Notation.lean]: Even without the notation for `ω`, `open Ordinal` is required.
And, in Lean 4, when an namespace is opened, its locale is automatically opened too, so `open_locale ordinal` and the comment is not needed anymore.

Comment (semorrison) [reply Mathlib/SetTheory/Ordinal/Notation.lean]: Thanks, sorry to be slow.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**